### ISSUE TITLE
foreshadow/lvi building error , memcmp running error

### ISSUE THREAD

**Issue by xiaonan-INTC** [jovanbulck/sgx-step#50]
1. Building error

For app/foreshadow/, run $make,
[===] Enclave [===]
[GEN] sgx_edger8r encl.edl
[CC]  encl_t.c (trusted edge)
[CC]  encl.c (core)
[LD]   encl.o encl_t.o -lsgx_trts encl.unsigned.so
[SGN] encl.unsigned.so
make[1]: *** [Makefile:53: encl.so] Error 255
make: *** [Makefile:73: build-Enclave] Error 2

For app/lvi/, run $make,
[===] Enclave [===]
[GEN] sgx_edger8r encl.edl
[CC]  encl_t.c (trusted edge)
[CC]  encl.c (core)
[AS]  asm.S (core)
[LD]   encl.o asm.o encl_t.o -lsgx_trts encl.unsigned.so
[SGN] encl.unsigned.so
make[1]: *** [Makefile:53: encl.so] Error 255
make: *** [Makefile:66: build-Enclave] Error 2

2. Running error

For memcmp:
[main.c] assertion 'pwd_len < MAX_LEN' failed: Invalid argument
Aborted

**Comment by xiaonan-INTC** [comment]
> 

Thanks a lot for the analysis, jovanbulck,

Update 1, 
lvi building: PASS, running: PASS

foreshadow building: PASS, running: FAIL, error message:
...
[main.c] prefetching enclave secret (EENTER/EEXIT)...
[main.c] extracting secret from L1 cache..
Illegal instruction
...

So I will close this issue and move all running errors to another issue https://github.com/jovanbulck/sgx-step/issues/51
For 2, 
memcmp building: PASS, running: FAIL

**Comment by namrata690** [comment]
Hello @xiaonan-INTC, 
I am getting a similar build error on running `make` in app/memcmp:


[===] Enclave [===]
[GEN] sgx_edger8r encl.edl
[CC] encl_t.c (trusted edge)
[CC] encl.c (core)
[LD] encl.o encl_t.o -lsgx_trts encl.unsigned.so
[SGN] encl.unsigned.so
make[1]: *** [Makefile:53: encl.so] Error 255
make: *** [Makefile:73: build-Enclave] Error 2

How did you resolve this?
Thanks.

**Comment by jovanbulck** [comment]
Hi @namrata690 

From above:

> the building errors should def not happen. Not sure what goes wrong exactly, you might want to comment the .SILENT in the Makefiles so you can intermediate output. I'd think it's something wrong w the installation on your end, since building works fine in the [CI](https://github.com/jovanbulck/sgx-step/actions/workflows/ci.yaml) (can't run the apps there but at least they should build without errors). let me know if you have more concrete issues w building?

I suspect your SDK installation is wrong. Can you post the full output of trying to build without the `--silent` flags in the Makefile?

https://github.com/jovanbulck/sgx-step/blob/1483dc898f288fcd4e7e1c328139d3f81c044c4d/app/memcmp/Makefile#L60

**Comment by namrata690** [comment]
> Hi @namrata690
> 
> From above:
> 
> > the building errors should def not happen. Not sure what goes wrong exactly, you might want to comment the .SILENT in the Makefiles so you can intermediate output. I'd think it's something wrong w the installation on your end, since building works fine in the [CI](https://github.com/jovanbulck/sgx-step/actions/workflows/ci.yaml) (can't run the apps there but at least they should build without errors). let me know if you have more concrete issues w building?
> 
> I suspect your SDK installation is wrong. Can you post the full output of trying to build without the `--silent` flags in the Makefile?
> 
> https://github.com/jovanbulck/sgx-step/blob/1483dc898f288fcd4e7e1c328139d3f81c044c4d/app/memcmp/Makefile#L60

![image](https://github.com/user-attachments/assets/a76a30c3-b509-4555-a9f1-ad74826aab89)

**Comment by jovanbulck** [comment]
Looks like your enclave signing is not succeeding. Try make clean first to generate new keys?

Also, try rmeoving the `> /dev/null 2> /dev/null` to see the signing error

**Comment by namrata690** [comment]
> Looks like your enclave signing is not succeeding. Try make clean first to generate new keys?
> 
> Also, try rmeoving the `> /dev/null 2> /dev/null` to see the signing error

![image](https://github.com/user-attachments/assets/88dd2db4-3165-4035-80cc-ea655a9f8dd5)

**Comment by jovanbulck** [comment]
looks like you'll have to remove and re-generate private_key.pem

**Comment by namrata690** [comment]
> looks like you'll have to remove and re-generate private_key.pem

Thanks, this worked! build succeeds now, but when I do `make run`, it shows the following:

![image](https://github.com/user-attachments/assets/11da3377-6049-4e5f-9b0a-63ee0c3439f5)

**Comment by jovanbulck** [comment]
glad to hear it worked!

the error message is exactly what it says, you have to load the driver /dev/sgx-step first -- see the README :)

**Comment by jovanbulck** [comment]
also note to self: improve the verbosity of the Makefile build system to display errors more clearly (eg via a global config option in the refactored build system cf #67 )

**Comment by namrata690** [comment]
> glad to hear it worked!
> 
> the error message is exactly what it says, you have to load the driver /dev/sgx-step first -- see the README :)

Thanks a lot, this worked, but the laptop froze while running, and I had to restart it again, known issue (or flaky behavior)?

![image](https://github.com/user-attachments/assets/1dd2e642-f02f-44a2-92a3-ec6997b60e84)

**Comment by xiaonan-INTC** [comment]
> > glad to hear it worked!
> > the error message is exactly what it says, you have to load the driver /dev/sgx-step first -- see the README :)
> 
> Thanks a lot, this worked, but the laptop froze while running, and I had to restart it again, known issue (or flaky behavior)?
> 
> ![image](https://private-user-images.githubusercontent.com/34477711/371144728-1dd2e642-f02f-44a2-92a3-ec6997b60e84.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MTc4OTksIm5iZiI6MTcyNzQxNzU5OSwicGF0aCI6Ii8zNDQ3NzcxMS8zNzExNDQ3MjgtMWRkMmU2NDItZjAyZi00NGEyLTkyYTMtZWM2OTk3YjYwZTg0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDA2MTMxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3OTY0NmFmZTgxNmVjZWY1NDQxM2I5ZGQzZDE3YWZmMWI2Yzg2Yzg5NGQ4NTA2YzU3YjY5NWUwMTZkZDkwOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.crbhhGQ5akfntE13KbKlmQPyM7YCR-7RvAOuEO9J7_o)

Also met same issue, could developers help check? :)

**Comment by jovanbulck** [comment]
Hi both!

From the screenshot above, it seems like your APIC setup is not working and I suspect you get a #GP which the kernel doesn't expect and freezes your machine..

FWIW, it seems it crashes already at this line when it tries to read from the APIC (using the new x2apic support) via a privileged IRQ gate:

https://github.com/jovanbulck/sgx-step/blob/1483dc898f288fcd4e7e1c328139d3f81c044c4d/libsgxstep/apic.c#L84

TL;DR: **first** make sure your basic setup works by running the programs in `app/selftest` before attempting app/memcmp or others. Importantly, make sure your kernel cmdline args are correct (cf README `sudo ./check_sys.sh`)

Feel free to let me know (preferably by opening a new issue with a much as info output etc as possible) if `app/selftest` doesn't work / freezes your system even  after `check_sys.sh` says it's correctly configured?

Hope this helps!

**Comment by jovanbulck** [comment]
also make sure to correctly configure `nox2apic` and `X2APIC` in config.h as described here: #89

**Comment by namrata690** [comment]
> also make sure to correctly configure `nox2apic` and `X2APIC` in config.h as described here: #89

Yeah, @w-woloszyn helped debug this and it works now. Thanks!
